distributed/tests/test_failed_workers.py

```diff
@@ -48,21 +49,57 @@ def test_submit_after_failed_worker_sync(loop):
             assert total.result() == sum(map(inc, range(10)))
 
 
-@pytest.mark.slow()
-@pytest.mark.parametrize("compute_on_failed", [False, True])
-@gen_cluster(client=True, config={"distributed.comm.timeouts.connect": "500ms"})
-async def test_submit_after_failed_worker_async(c, s, a, b, compute_on_failed):
-    async with Nanny(s.address, nthreads=2) as n:
-        await c.wait_for_workers(3)
+@pytest.mark.repeat(20)  # TEMP DO NOT MERGE
+@pytest.mark.slow
+@pytest.mark.parametrize("wait_closing", [False, True])
+@pytest.mark.parametrize("y_on_failed", [False, True])
+@pytest.mark.parametrize("x_on_failed", [False, True])
+@gen_cluster(client=True, nthreads=[("", 1)] * 2)
+async def test_submit_after_failed_worker_async(
+    c, s, a, b, x_on_failed, y_on_failed, wait_closing
+):
+    a_ws = s.workers[a.address]
+    b_ws = s.workers[b.address]
+
+    L = c.map(
+        inc,
+        range(10),
+        workers=[b.address if x_on_failed else a.address],
+        allow_other_workers=True,
+    )
 
-        L = c.map(inc, range(10))
-        await wait(L)
+    await wait(L)
+    total = c.submit(
+        sum,
+        L,
+        key="y",
+        workers=[b.address if y_on_failed else a.address],
+        allow_other_workers=True,
+    )
+
+    done_update_graph = False
+    if wait_closing:
+        in_update_graph = asyncio.Event()
+
+        async def update_graph(*args, **kwargs):
+            in_update_graph.set()
+            await async_poll_for(
+                lambda: b_ws.status == Status.closing, timeout=5, period=0
+            )
+            s.update_graph(*args, **kwargs)
+            nonlocal done_update_graph
+            done_update_graph = True
+
+        s.stream_handlers["update-graph"] = update_graph
+        await in_update_graph.wait()
```

Comment:
Given the problem, I would've expected something like


```python

def block(arg, enter, exit):
    enter.set()
    exit.wait()
    return arg

enter = Event()
exit = Event()
enter2 = Event()
exit2 = Event()

d1 = c.submit(inc, 0, key='d1', workers=["A"])
d2 = c.submit(block, 1, enter=enter, exit=exit, key='d2', workers=["B"])

x = c.submit(sum, [d1, d2], key='x')
block_executor = c.submit(block, None, enter=enter2, exit=exit2, key='x', workers=["B"])

await enter.wait()
await enter2.wait()
await asyncio.gather([
    exit.set(),
    B.close()
])
```

I haven't tested the above and it may still need fine tuning but I would expect something like this to trigger the condition you're talking about. d2 completes while B is closing s.t. when d2 finishes, x is transitioned while B is still closing.

I don't entirely understand why we need update_graph to trigger this condition

Reply:
Especially considering that update_graph (right now) is still synchronous.

Reply:
I'm not too happy with my test, either. But I don't think your suggestion works (read below).

What we're trying to test:
- `decide_worker_non_rootish(ts)` is called on a task with `workers=[b.address], allow_other_workers=True`,
- with the dependencies of the task partially on a and partially on b,
- while b is in `closing` status, and 
- with `decide_worker(ts, valid_workers=set(), all_workers={a})` that would pick b from `ts.dependencies[...].who_has`, due to less dependency bytes needing to be transferred to it, but instead picks a because b is not in all_workers.

The problem with writing the test is that we need to time `update_graph` to land *exactly* during the 1-2 event loop cycles while the worker is in `closing` status.
The worker transitions from `closing` to being removed when the batched comms collapse, here:
https://github.com/dask/distributed/blob/f0303aaff3a22fd00660312cfa3f74a1fa178938/distributed/scheduler.py#L5700-L5705

Alternatively to monkey-patching update-graph, I could have
- monkey-patched Scheduler.remove_worker. In hindsight that's a better idea; I'll have a look at it now.
- synchronously call update_graph directly on the scheduler and update the state on the client by hand (complicated and brittle).

## In your code:
I think that the scheduler will never receive the task-finished message for d2, since it's a whole 2 cycles of event loop after Worker.close collapses the batched comms.

I'm not sure why you think the scheduler should receive `{op: task-finished, key: d2}` deterministically after `{op: worker-status-change, worker: <b>, status: closing}`, but deterministically before the collapse of the TCP channel?

Also I can't understand what's the purpose of `block_executor`.

Reply:
>  but deterministically before the collapse of the TCP channel?

I never said it would do so deterministically. I said I would expect _something like this_ to trigger the condition you described. I also said it would require more fine tuning like distributing nbytes properly and maybe introduce an event somewhere.